File: app/views/api/incidents.ts
export const INCIDENT_TYPES = [
  'Flood',
  'Fire',
  'Medical',
  'Storm',
  'Shelter',
  'Relief',
  'Infrastructure',
  'Search and Rescue',
] as const;

export type IncidentType = typeof INCIDENT_TYPES[number];

export interface Incident {
  id: number;
  title: string;
  description: string;
  type: string;
  latitude: number;
  longitude: number;
  priority: 'low' | 'normal' | 'high' | 'critical';
  status: 'open' | 'closed';
  requiredCertificate: string | null;
  availableAt: string;
  createdBy: number;
}

export interface ActiveVolunteerIncident extends Incident {
  joinedAt: string;
  participationStatus: 'joined';
}

export interface CreateIncidentInput {
  title: string;
  description: string;
  incident_type: IncidentType;
  important_data: string;
  important_data_extra: string;
  latitude: number;
  longitude: number;
  priority: Incident['priority'];
  required_certificate: string | null;
  available_at: string | null;
}

const API_BASE_URL = (import.meta.env.VITE_API_BASE_URL || '').replace(/\/$/, '');

const TYPE_COLORS: Record<string, string> = {
  flood: '#ef4444',
  fire: '#f97316',
  medical: '#ec4899',
  relief: '#f59e0b',
  shelter: '#22c55e',
  storm: '#3b82f6',
  infrastructure: '#a855f7',
  'search and rescue': '#2EC4B6',
};

export function getIncidentTypeColor(type: string): string {
  const normalizedType = type.trim().toLowerCase();
  if (TYPE_COLORS[normalizedType]) return TYPE_COLORS[normalizedType];

  const fallbackColors = ['#2EC4B6', '#a855f7', '#ec4899', '#14b8a6', '#eab308'];
  const colorIndex = Array.from(normalizedType).reduce(
    (total, character) => total + character.charCodeAt(0),
    0,
  ) % fallbackColors.length;

  return fallbackColors[colorIndex];
}

function normalizePriority(priority: string): Incident['priority'] {
  const value = priority.toLowerCase();

  if (value === 'urgent') return 'critical';
  if (value === 'medium') return 'normal';
  if (value === 'low' || value === 'normal' || value === 'high' || value === 'critical') return value;

  return 'normal';
}

function normalizeIncident(raw: Record<string, unknown>): Incident {
  const endedAt = raw.Ended_at ?? raw.endedAt;
  const availableAt = raw.availableAt ?? raw.Available_at ?? raw.createdAt ?? raw.Created_at;

  return {
    id: Number(raw.Incident_id ?? raw.incidentId ?? raw.id),
    title: String(raw.Title ?? raw.title ?? 'Untitled incident'),
    description: String(raw.Description ?? raw.description ?? ''),
    type: String(raw.Type ?? raw.type ?? 'Other'),
    latitude: Number(raw.Latitude ?? raw.latitude),
    longitude: Number(raw.Longitude ?? raw.longitude),
    priority: normalizePriority(String(raw.Priority ?? raw.priority ?? 'normal')),
    status: endedAt ? 'closed' : String(raw.status ?? 'open').toLowerCase() === 'closed' ? 'closed' : 'open',
    requiredCertificate: raw.requiredCertificate
      ? String(raw.requiredCertificate)
      : null,
    availableAt: String(availableAt),
    createdBy: Number(raw.createdBy ?? raw.Created_by ?? 0),
  };
}

export function isIncidentUpcoming(incident: Incident, now = Date.now()): boolean {
  return new Date(incident.availableAt).getTime() > now;
}

export function formatIncidentCountdown(availableAt: string, now = Date.now()): string {
  const remainingSeconds = Math.max(
    0,
    Math.ceil((new Date(availableAt).getTime() - now) / 1000),
  );
  if (remainingSeconds === 0) return 'Available now';

  const days = Math.floor(remainingSeconds / 86400);
  const hours = Math.floor((remainingSeconds % 86400) / 3600);
  const minutes = Math.floor((remainingSeconds % 3600) / 60);
  const seconds = remainingSeconds % 60;

  if (days > 0) return `${days}d ${hours}h ${minutes}m`;
  if (hours > 0) return `${hours}h ${minutes}m ${seconds}s`;
  if (minutes > 0) return `${minutes}m ${seconds}s`;
  return `${seconds}s`;
}

async function request<T>(path: string, options?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE_URL}${path}`, {
    headers: { 'Content-Type': 'application/json' },
    ...options,
  });

  if (!response.ok) {
    const payload = await response.json().catch(() => null);
    throw new Error(payload?.detail || `Request failed with status ${response.status}`);
  }

  return response.json() as Promise<T>;
}

export async function createIncident(input: CreateIncidentInput): Promise<number> {
  const user = JSON.parse(localStorage.getItem('plymouth-user') ?? 'null') as {
    token?: string;
  } | null;
  if (!user?.token) throw new Error('Your session is missing. Please log in again.');

  const result = await request<{ incidentId: number }>('/incidents', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${user.token}`,
    },
    body: JSON.stringify(input),
  });
  return result.incidentId;
}

export async function getIncidents(): Promise<Incident[]> {
  const data = await request<unknown[]>('/incidents');
  return data
    .map(item => normalizeIncident(item as Record<string, unknown>))
    .filter(incident => incident.status === 'open');
}

export async function getIncident(id: number): Promise<Incident | null> {
  try {
    const data = await request<Record<string, unknown>>(`/incidents/${id}`);
    return normalizeIncident(data);
  } catch {
    return null;
  }
}

export async function getMyActiveIncidents(): Promise<ActiveVolunteerIncident[]> {
  const user = JSON.parse(localStorage.getItem('plymouth-user') ?? 'null') as {
    token?: string;
  } | null;
  if (!user?.token) throw new Error('Your session is missing. Please log in again.');

  const data = await request<Record<string, unknown>[]>('/incidents/mine/active', {
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${user.token}`,
    },
  });

  return data.map(item => ({
    ...normalizeIncident(item),
    joinedAt: String(item.joinedAt),
    participationStatus: 'joined' as const,
  }));
}

export async function volunteerForIncident(incidentId: number): Promise<void> {
  const user = JSON.parse(localStorage.getItem('plymouth-user') ?? 'null') as {
    token?: string;
  } | null;
  if (!user?.token) throw new Error('Your session is missing. Please log in again.');

  await request(`/incidents/${incidentId}/join`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${user.token}`,
    },
  });
}

export async function closeIncident(incidentId: number): Promise<void> {
  const user = JSON.parse(localStorage.getItem('plymouth-user') ?? 'null') as {
    token?: string;
  } | null;
  if (!user?.token) throw new Error('Your session is missing. Please log in again.');

  await request(`/incidents/${incidentId}/close`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${user.token}`,
    },
  });
}

File: app/routes/Incidents.py
from datetime import datetime, timezone
from fastapi import APIRouter, Header, HTTPException
from pydantic import BaseModel, Field
from typing import Literal
from decimal import Decimal
from app.repositories.coordinator_repository import CoordinatorRepository
from app.repositories.user_account import UserAccount
from app.security import verify_access_token



service = CoordinatorRepository()
user_service = UserAccount()
router = APIRouter()

class Incident(BaseModel):
    title: str = Field(min_length=3, max_length=255)
    description: str = Field(min_length=10)
    incident_type: Literal[
        "Flood",
        "Fire",
        "Medical",
        "Storm",
        "Shelter",
        "Relief",
        "Infrastructure",
        "Search and Rescue",
    ]
    important_data: str = Field(max_length=255)
    important_data_extra: str = Field(max_length=255)
    latitude: Decimal = Field(ge=Decimal("50.32"), le=Decimal("50.45"))
    longitude: Decimal = Field(ge=Decimal("-4.25"), le=Decimal("-4.02"))
    priority: Literal["low", "normal", "high", "critical"]
    required_certificate: Literal[
        "First Aid",
        "Water Rescue",
        "Safeguarding",
        "Manual Handling",
        "Working at Height",
    ] | None = None
    status: bool = True
    created_at : datetime | None = None
    available_at: datetime | None = None
    created_by : int | None = None
    ended_at : datetime | None = None
    ended_by : int | None = None


class priority_update(BaseModel):
    priority: str | None = None

class update_des(BaseModel):
    description: str | None = None

def serialize_incident(incident: dict):
    available_at = incident["availableAt"]
    if available_at is not None and available_at.tzinfo is None:
        available_at = available_at.replace(tzinfo=timezone.utc)

    result = {
        "id": incident["incidentId"],
        "incidentId": incident["incidentId"],
        "title": incident["title"],
        "description": incident["description"],
        "type": incident["type"],
        "latitude": float(incident["latitude"]),
        "longitude": float(incident["longitude"]),
        "priority": incident["priority"],
        "requiredCertificate": incident["requiredCertificate"],
        "status": "closed" if incident["endedAt"] else "open",
        "createdAt": incident["createdAt"],
        "availableAt": available_at,
        "createdBy": incident["createdBy"],
    }
    if "joinedAt" in incident:
        result["joinedAt"] = incident["joinedAt"]
        result["participationStatus"] = incident["participationStatus"]
    return result


@router.get("/incidents")
def get_incidents():
    try:
        return [
            serialize_incident(incident)
            for incident in service.list_active_incidents()
        ]
    except Exception as error:
        raise HTTPException(500, "Failed to load incidents") from error


@router.get("/incidents/mine/active")
def get_my_active_incidents(authorization: str | None = Header(default=None)):
    if not authorization or not authorization.startswith("Bearer "):
        raise HTTPException(401, "Authentication is required")

    user_id = verify_access_token(authorization.removeprefix("Bearer ").strip())
    if user_id is None:
        raise HTTPException(401, "Your session is invalid or has expired")
    if user_service.get_user_role(user_id) != "volunteer":
        raise HTTPException(403, "Only volunteers have joined task lists")

    try:
        return [
            serialize_incident(incident)
            for incident in user_service.list_active_incidents_for_volunteer(user_id)
        ]
    except Exception as error:
        raise HTTPException(500, "Failed to load your active tasks") from error


@router.get("/incidents/{incident_id}")
def get_incident(incident_id: int):
    try:
        incident = service.get_incident(incident_id)
    except Exception as error:
        raise HTTPException(500, "Failed to load incident") from error

    if incident is None:
        raise HTTPException(404, "Incident not found")

    return serialize_incident(incident)

#create
#FIXME: convert location to coords method has no endpoint yet.
@router.post("/incidents", status_code=201)
def create_incident(incident: Incident, authorization: str | None = Header(default=None)):
    if not authorization or not authorization.startswith("Bearer "):
        raise HTTPException(401, "Authentication is required")

    coordinator_id = verify_access_token(authorization.removeprefix("Bearer ").strip())
    if coordinator_id is None:
        raise HTTPException(401, "Your session is invalid or has expired")

    if not user_service.can_create_incidents(coordinator_id):
        raise HTTPException(403, "Only an active coordinator can create incidents")

    incident_id = service.create_incident(incident, coordinator_id)
    if incident_id is None:
        raise HTTPException(500, "Failed to create incident")

    return {
        "success": True,
        "message": "Incident created",
        "incidentId": incident_id,
    }

#update description
@router.patch("/incidents/{incident_id}")
def update_incident_description(incident_id: int, update: update_des):
    if update.description is None:
        raise HTTPException(400, "Description is required")
    
    success = service.update_description(incident_id , update.description)
    if not success:
        raise HTTPException(400, "Incident not found")
    
    return {"success": True, "message": "Description updated successfully"}
 
#update priority
@router.patch("/incidents/{incident_id}/priority")
def update_incident_priority(incident_id: int, update: priority_update):
    if update.priority is None:
        raise HTTPException(400, "Priority is required")

    success = service.update_priority(incident_id, update.priority)
    if not success:
        raise HTTPException(400, "Incident not found")
    
    return {"success": True, "message": "Priority updated successfully"}


@router.post("/incidents/{incident_id}/join")
def join_incident(
    incident_id: int,
    authorization: str | None = Header(default=None),
):
    if not authorization or not authorization.startswith("Bearer "):
        raise HTTPException(401, "Authentication is required")

    user_id = verify_access_token(authorization.removeprefix("Bearer ").strip())
    if user_id is None:
        raise HTTPException(401, "Your session is invalid or has expired")
    if user_service.get_user_role(user_id) != "volunteer":
        raise HTTPException(403, "Only volunteers can join incidents")

    incident = service.get_incident(incident_id)
    if incident is None:
        raise HTTPException(404, "Incident not found")
    if incident["endedAt"] is not None or not incident["status"]:
        raise HTTPException(409, "This incident has ended")
    available_at = incident["availableAt"]
    if available_at is not None and available_at > datetime.now(timezone.utc).replace(tzinfo=None):
        raise HTTPException(409, "This incident is not available yet")

    required_certificate = incident["requiredCertificate"]
    if (
        required_certificate
        and not user_service.has_verified_certificate(user_id, required_certificate)
    ):
        raise HTTPException(
            403,
            f"This incident requires a verified {required_certificate} certificate",
        )

    if not user_service.volunteer_for(user_id, incident_id):
        raise HTTPException(409, "Unable to volunteer for this incident")

    return {"success": True, "message": "Volunteered successfully"}


@router.post("/incidents/{incident_id}/close")
def close_incident(
    incident_id: int,
    authorization: str | None = Header(default=None),
):
    if not authorization or not authorization.startswith("Bearer "):
        raise HTTPException(401, "Authentication is required")

    user_id = verify_access_token(authorization.removeprefix("Bearer ").strip())
    if user_id is None:
        raise HTTPException(401, "Your session is invalid or has expired")

    role = user_service.get_user_role(user_id)
    if role not in {"coordinator", "system_manager"}:
        raise HTTPException(
            403,
            "Only coordinators and system managers can end incidents",
        )

    incident = service.get_incident(incident_id)
    if incident is None:
        raise HTTPException(404, "Incident not found")
    if incident["endedAt"] is not None or not incident["status"]:
        raise HTTPException(409, "Incident has already ended")
    if role == "coordinator" and incident["createdBy"] != user_id:
        raise HTTPException(403, "Coordinators can only end incidents they created")

    try:
        closed = service.close_incident(
            user_id,
            incident_id,
            allow_any_incident=role == "system_manager",
        )
    except Exception as error:
        raise HTTPException(500, "Failed to end incident") from error

    if not closed:
        raise HTTPException(409, "Incident could not be ended")

    return {"success": True, "message": "Incident ended"}
